### ISSUE TITLE
`torch.func.jvp` yields different output compared to original function with `torch.nn.functional.layer_norm`

### ISSUE THREAD

**Issue by lengstrom** [pytorch/pytorch#124533]
### 🐛 Describe the bug

When I calculate a hessian vector product of layernorm with `torch.func.jvp`, the returned output from `jvp` (i.e., the value of the function, or the average gradient of the layernorm output) evaluated at the primals is different compared to calling the function directly. Code is below; I add the beginning part to ensure that the code is deterministic. In contrast, a linear layer has no difference in outputs (code for both linear layer and layernorm below).

```
import os
os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":4096:8"
import torch as ch
import numpy as np
import random

def set_seed(seed):
    os.environ['PYTHONHASHSEED'] = str(seed)
    ch.manual_seed(seed)
    random.seed(seed)
    np.random.seed(seed)
    np.random.default_rng(seed=seed)

    ch.cuda.manual_seed(seed)
    ch.cuda.manual_seed_all(seed)
    ch.backends.cudnn.deterministic = True

    ch.backends.cudnn.benchmark = False
    ch.backends.cuda.matmul.allow_tf32 = True # TODO: check
    ch.backends.cudnn.allow_tf32 = True # TODO: check
    ch.backends.cudnn.enabled = True

    ch.backends.cuda.enable_flash_sdp(False)
    ch.backends.cuda.enable_mem_efficient_sdp(False)
    ch.backends.cuda.enable_math_sdp(True)

set_seed(0)

def test_fn(primals, fn, x):
    def grads(*params):
        def g(params):
            y = fn(x, *params)
            return y.sum()

        return ch.func.grad(g)(params)

    tangents = tuple(ch.randn_like(p) for p in primals)

    output = grads(*primals)
    jvp_output, _ = ch.func.jvp(grads, primals, tangents)

    if not ch.is_tensor(output):
        output = ch.cat([o.flatten() for o in output], 0)
        jvp_output = ch.cat([o.flatten() for o in jvp_output], 0)

    print('norm(output - jvp_output)', (output - jvp_output).norm())


### TEST LAYERNORM
params = (ch.randn(20).cuda(), ch.randn(20).cuda())
x = ch.randn(100, 20).cuda()

def layer_norm(x, ln_weight, ln_bias):
    return ch.nn.functional.layer_norm(x, (20,), weight=ln_weight, bias=ln_bias)

print('LAYERNORM')
test_fn(params, layer_norm, x)

### TEST LINEAR
params = (ch.randn(20, 20).cuda(),)
def linear(x, w):
    return ch.nn.functional.linear(x, w)

print('LINEAR')
test_fn(params, linear, x)
```
has output:
```
LAYERNORM
norm(output - jvp_output) tensor(6.5185e-06, device='cuda:0')
LINEAR
norm(output - jvp_output) tensor(0., device='cuda:0')
```

### Versions

```
PyTorch version: 2.4.0.dev20240409+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.26.3
Libc version: glibc-2.35

Python version: 3.10.13 (main, Sep 11 2023, 13:44:35) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-100-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA A100 80GB PCIe
GPU 1: NVIDIA A100 80GB PCIe
GPU 2: NVIDIA A100 80GB PCIe
GPU 3: NVIDIA A100 80GB PCIe
GPU 4: NVIDIA A100 80GB PCIe
GPU 5: NVIDIA A100 80GB PCIe
GPU 6: NVIDIA A100 80GB PCIe
GPU 7: NVIDIA A100 80GB PCIe

Nvidia driver version: 535.54.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      48 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             128
On-line CPU(s) list:                0-127
Vendor ID:                          AuthenticAMD
Model name:                         AMD EPYC 7513 32-Core Processor
CPU family:                         25
Model:                              1
Thread(s) per core:                 2
Core(s) per socket:                 32
Socket(s):                          2
Stepping:                           1
Frequency boost:                    enabled
CPU max MHz:                        3681.6399
CPU min MHz:                        1500.0000
BogoMIPS:                           5190.19
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca
Virtualization:                     AMD-V
L1d cache:                          2 MiB (64 instances)
L1i cache:                          2 MiB (64 instances)
L2 cache:                           32 MiB (64 instances)
L3 cache:                           256 MiB (8 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0-31,64-95
NUMA node1 CPU(s):                  32-63,96-127
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Mitigation; safe RET
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] onnx==1.14.0
[pip3] onnxruntime==1.15.1
[pip3] optree==0.11.0
[pip3] pytorch-pfn-extras==0.7.6
[pip3] pytorch-ranger==0.1.1
[pip3] pytorch-triton==3.0.0+989adb9a29
[pip3] torch==2.4.0.dev20240409+cu121
[pip3] torch-optimizer==0.3.0
[pip3] torchaudio==2.2.0.dev20240409+cu121
[pip3] torchmetrics==1.3.1
[pip3] torchopt==0.7.3
[pip3] torchvision==0.19.0.dev20240409+cu121
[pip3] triton==2.2.0
[pip3] triton-pre-mlir==2.0.0
[conda] cudatoolkit               10.1.243             h6bb024c_0
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] optree                    0.11.0                   pypi_0    pypi
[conda] pytorch-pfn-extras        0.7.6                    pypi_0    pypi
[conda] pytorch-ranger            0.1.1                    pypi_0    pypi
[conda] pytorch-triton            3.0.0+989adb9a29          pypi_0    pypi
[conda] torch                     2.4.0.dev20240409+cu121          pypi_0    pypi
[conda] torch-optimizer           0.3.0                    pypi_0    pypi
[conda] torchaudio                2.2.0.dev20240409+cu121          pypi_0    pypi
[conda] torchmetrics              1.3.1                    pypi_0    pypi
[conda] torchopt                  0.7.3                    pypi_0    pypi
[conda] torchvision               0.19.0.dev20240409+cu121          pypi_0    pypi
[conda] triton                    2.2.0                    pypi_0    pypi
[conda] triton-pre-mlir           2.0.0                    pypi_0    pypi
```

cc @zou3519 @Chillee @samdow @kshitij12345 @janeyx99

**Comment by zou3519** [comment]
This seems within fp precision tolerances, so this doesn't look like a bug.

**Comment by zou3519** [comment]
Using float64 results in 1e-14 difference, so yeah, this is expected behavior.
